applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,741 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <assert.h>
+#include <stdio.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+#include <device.h>
+#include <drivers/uart.h>
+#include <misc/byteorder.h>
+
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+
+#include "ble_controller_hci_vs.h"
+
+#include "chmap_filter.h"
+
+#define MODULE ble_qos
+#include "module_state_event.h"
+#include "ble_event.h"
+#include "config_event.h"
+#include "hid_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_BLE_QOS_LOG_LEVEL);
+
+#define INVALID_BLACKLIST 0xFFFF
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+#define THREAD_STACK_SIZE \
+	(CONFIG_DESKTOP_BLE_QOS_STACK_SIZE + \
+	CONFIG_DESKTOP_BLE_QOS_STATS_PRINT_STACK_SIZE)
+#else
+#define THREAD_STACK_SIZE CONFIG_DESKTOP_BLE_QOS_STACK_SIZE
+#endif
+
+#if CONFIG_DESKTOP_BLE_QOS_STATS_PRINTOUT_ENABLE
+# if DT_NORDIC_NRF_USBD_USBD_0_NUM_IN_ENDPOINTS < 4
+# error Too few USB IN Endpoints enabled.\
+Modify appropriate dts.overlay to increase num-in-endpoints to 4 or more
+# endif
+#endif
+
+#ifndef ROUNDED_DIV
+#define ROUNDED_DIV(a, b)  (((a) + ((b) / 2)) / (b))
+#endif /* ROUNDED_DIV */
+
+/* Lowest thread priority */
+#define THREAD_PRIORITY K_PRIO_PREEMPT((CONFIG_NUM_PREEMPT_PRIORITIES - 1))
+
+static K_THREAD_STACK_DEFINE(thread_stack, THREAD_STACK_SIZE);
+static struct k_thread thread;
+
+struct params_ble {
+	u16_t sample_count_min;
+	u8_t min_channel_count;
+	s16_t weight_crc_ok;
+	s16_t weight_crc_error;
+	u16_t ble_block_threshold;
+	u8_t eval_max_count;
+	u16_t eval_duration;
+	u16_t eval_keepout_duration;
+	u16_t eval_success_threshold;
+} __packed;
```

Comment:
packed is not needed anymore since data is explicitly deserialized

Reply:
The size of the struct will change in this case, and can no longer be used for sizeof() allocation of fetch events, unless the struct members are reorganized

Reply:
I don't think it should be used this way (seems like an abuse of packed). It would be safer to calculate sizeof the packet using sizes of members.
This said I understand this this makes things easier so unless somebody else complains let's leave it.